src/middlewared/middlewared/plugins/activedirectory.py

```diff
@@ -0,0 +1,1147 @@
+        servers = []
+        if await self.middleware.call('cache.has_key', f'SRVCACHE_{srv.name}_SITE'):
+            servers = await self.middleware.call('cache.get', f'SRVCACHE_{srv.name}_SITE')
+
+        if not servers and await self.middleware.call('cache.has_key', f'SRVCACHE_{srv.name}'):
+            servers = await self.middleware.call('cache.get', f'SRVCACHE_{srv.name}')
+
+        return servers
+
+    @private
+    async def _set_cached_srv_records(self, srv=None, site=None, results=[]):
+        """
+        Cache srv record lookups for 24 hours
+        """
+        if not srv:
+            raise CallError('srv record type not specified', errno.EINVAL)
+
+        if site:
+            await self.middleware.call('cache.put', f'SRVCACHE_{srv.name}_SITE', results, 86400)
+        else:
+            await self.middleware.call('cache.put', f'SRVCACHE_{srv.name}', results, 86400)
+        return True
+
+    @accepts()
+    async def started(self):
+        """
+        Issue a no-effect command to our DC. This checks if our secure channel connection to our
+        domain controller is still alive. It has much less impact than wbinfo -t.
+        Default winbind request timeout is 60 seconds, and can be adjusted by the smb4.conf parameter
+        'winbind request timeout ='
+        """
+        netlogon_ping = await run(['wbinfo', '-P'], check=False)
+        if netlogon_ping.returncode != 0:
+            await self._set_state(DSStatus['FAULTED'])
+            return False
+        await self._set_state(DSStatus['HEALTHY'])
+        return True
+
+    @private
+    async def _net_ads_join(self):
+        ad = await self.config()
+        if ad['createcomputer']:
+            netads = await run([
+                'net', '-k', '-U', ad['bindname'], '-d', '5',
+                'ads', 'join', f'createcomputer={ad["createcomputer"]}', 
+                ad['domainname'],], check=False)
+        else:
+            netads = await run([
+                'net', '-k', '-U', ad['bindname'], '-d', '5',
+                'ads', 'join', ad['domainname'],], check=False)
+
+        if netads.returncode != 0:
+            await self._set_state(DSStatus['FAULTED'])
+            raise CallError(f'Failed to join [{ad["domainname"]}]: [{netads.stdout.decode().strip()}]')
+
+    @private
+    async def _net_ads_testjoin(self, workgroup):
+        ad = await self.config()
+        netads = await run([
+            'net', '-k', '-w', workgroup,
+            '-d', '5', 'ads', 'testjoin', ad['domainname'],
+            ],
+            check=False
+        )
+        if netads.returncode != 0:
+            errout = netads.stderr.decode().strip()
+            self.logger.debug(f'net ads testjoin failed with error: [{errout}]')
+            if '0xfffffff6' in errout: 
+                return neterr.NOTJOINED
+            else:
+                return neterr.FAULT
+
+        return neterr.JOINED
+
+    @private
+    async def get_netbios_domain_name(self):
```

Comment:
I think this should also be a method

Reply:
You may have missed to turn this into a method? Since there are sync calls being done here